cloud-controller-manager/do/firewall_controller.go

```diff
@@ -158,18 +171,37 @@ func NewFirewallController(
 
 // Run starts the firewall controller loop.
 func (fc *FirewallController) Run(ctx context.Context, stopCh <-chan struct{}, fwReconcileFrequency time.Duration) {
-	wait.Until(func() {
-		err := fc.observeRunLoopDuration(ctx, stopCh, fwReconcileFrequency)
-		if err != nil {
-			klog.Errorf("failed to run firewall controller loop: %v", err)
+	observeRunLoopDurationErr := fc.observeRunLoopDuration(ctx)
+	go wait.Until(func() {
+		// process next item
+		for fc.processNextItemAndCheckError(observeRunLoopDurationErr) {
 		}
 	}, fwReconcileFrequency, stopCh)
+
+	<-stopCh
+	klog.Info("Stopping firewall controller")
 }
 
-func (fc *FirewallController) actualRun(stopCh <-chan struct{}, fwReconcileFrequency time.Duration) error {
-	ctx, cancel := context.WithTimeout(context.Background(), fwReconcileFrequency)
-	defer cancel()
+func (fc *FirewallController) processNextItemAndCheckError(observeRunLoopDurationErr error) bool {
```

Comment:
I'm not sure the implementation we have here makes sense. The idea is to pick an item off of the queue and process it, but we do not seem to work with the `key` at all.